packages/router/router.ts

```diff
@@ -3231,7 +3234,8 @@ function getMatchesToLoad(
       nextParams: matches[matches.length - 1].params,
       ...submission,
       actionResult,
-      defaultShouldRevalidate,
+      // Forced revalidation due to submission, useRevalidate, or X-Remix-Revalidate
+      defaultShouldRevalidate: isRevalidationRequired,
```

Comment:
This is the crux of the bug fix.  Fetchers should not use the same logic as routes (which takes into account search params and same-URL navigations).  They should only revalidate on `action` submissions, `router.revalidate`, or `X-Remix-Revalidate`